sqlglot/optimizer/merge_subqueries.py

```diff
@@ -174,6 +174,22 @@ def _outer_select_joins_on_inner_select_join():
             for col in inner_projections[selection].find_all(exp.Column)
         )
 
+    def _is_recursive():
+        # Recursive CTEs look like this:
+        #     WITH RECURSIVE cte AS (
+        #       SELECT * FROM x  <-- inner scope
+        #       UNION ALL
+        #       SELECT * FROM cte  <-- outer scope
+        #     )
+        cte = inner_scope.expression.parent
+        node = outer_scope.expression.parent
+
+        while node:
+            if node is cte:
+                return True
```

Comment:
Shouldn't we check that the `recursive` flag is set as well?

Reply:
`recursive` is set on `With`, not `CTE`. We _can_, but I don't know if its necessary

Reply:
I guess I was a little confused about this loop/check being used as a proxy for detecting recursion because of the `_is_recursive` name. Is it sufficient? You can also have a CTE which is non-recursive and contains a `UNION` inside of it.

Reply:
IIUC, a recursive WITH statement can include non-recursive ctes _and_ recursive ctes, so you might still want to merge.
If the outer scope is a child of the inner scope, you know you're recursive. That should be true for all scopes.

Reply:
I see, makes sense